google/cloud/storage/internal/grpc_client_test.cc

```diff
@@ -0,0 +1,282 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/grpc_client.h"
+#include "google/cloud/storage/testing/mock_storage_stub.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include "google/cloud/testing_util/validate_metadata.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+namespace {
+
+namespace v2 = ::google::storage::v2;
+using ::google::cloud::testing_util::GetMetadata;
+using ::google::cloud::testing_util::StatusIs;
+using ::testing::HasSubstr;
+using ::testing::Pair;
+using ::testing::Return;
+using ::testing::UnorderedElementsAre;
+
+Status PermanentError() {
+  return Status(StatusCode::kPermissionDenied, "uh-oh");
+}
+
+TEST(GrpcClient, QueryResumableUpload) {
+  auto mock = std::make_shared<testing::MockStorageStub>();
+  EXPECT_CALL(*mock, QueryWriteStatus)
+      .WillOnce([](grpc::ClientContext& context,
+                   v2::QueryWriteStatusRequest const& request) {
+        auto metadata = GetMetadata(context);
+        EXPECT_THAT(metadata,
+                    UnorderedElementsAre(
+                        Pair("x-goog-quota-user", "test-quota-user"),
+                        // Map JSON names to the `resource` subobject
+                        Pair("x-goog-fieldmask", "resource(field1,field2)")));
+        EXPECT_EQ(request.upload_id(), "test-only-upload-id");
+        return PermanentError();
+      });
+  auto client = GrpcClient::CreateMock(mock);
+  auto status = client->QueryResumableUpload(
+      QueryResumableUploadRequest("test-only-upload-id")
+          .set_multiple_options(Fields("field1,field2"),
+                                QuotaUser("test-quota-user")));
+  EXPECT_THAT(status,
+              StatusIs(StatusCode::kPermissionDenied, HasSubstr("uh-oh")));
```

Comment:
Q.  Maybe these expectations should be written in terms of `PermanentError()`?

Reply:
Done.